hutch_python/ipython_session_timer.py

```diff
@@ -36,41 +36,62 @@ class IPythonSessionTimer:
 
     Attributes
     ----------
+    curr_time: float
+        The current time in seconds.
+
     max_idle_time: int
         The maximum number of seconds a user session can be idle (currently set
         to 172800 seconds or 48 hours).
+
+    last_active_time: float
+        The time of the last user activity in this session.
+
+    idle_time: float
+        The amount of time the user session has been idle.
     '''
 
     def __init__(self, ipython):
         global max_idle_time
-        self.max_idle_time = max_idle_time
+        self.curr_time = 0
+        self.max_idle_time = 30  # max_idle_time
+        self.last_active_time = 0
+        self.idle_time = 0
         self.user_active = False
 
-        # _set_last_active_time() function will trigger every time user runs a cell
         ipython.events.register('pre_run_cell', self._set_user_active)
         ipython.events.register('post_run_cell', self._set_user_inactive)
 
-    def _set_user_active(self, result):
+    def _set_user_active(self):
         self.user_active = True
+        self.last_active_time = time.time()
+        print("pre_cell_run")
 
-    def _set_user_inactive(self, result):
+    def _set_user_inactive(self):
         self.user_active = False
+        self.last_active_time = time.time()
+        print("post_cell_run")
+
+    def _get_time_passed(self):
+        self.curr_time = time.time()
+        self.idle_time = self.curr_time - self.last_active_time
+
+    def _timer(self, sleep_time):
+        time.sleep(sleep_time)
 
     def _start_session(self):
-        # poll for user activity
-        while (1):
-            if not self.user_active:
-                time.sleep(self.max_idle_time)
-                if not self.user_active:
-                    break
-            else:
-                # check if the user has become inactive once every minute
-                time.sleep(60)
-
-        # Close the user session
-        print("This hutch-python session has timed out and automatically closed. Please start a new session")
-
-        # Close this ipython session
+        # Check if idle_time has exceeded max_idle_time
+        while (self.idle_time < self.max_idle_time):
+
+            while (self.user_active):
+                time.sleep(6)
```

Comment:
Is there special significance to this number?  It's fine how it is, it just seems strangely specific

Reply:
It was for testing. It's supposed to check if the user is active once a minute but the test timeout duration was 20 seconds so I changed it to 6. It's been changed back to 60 seconds.